### PR TITLE
Add NNCLR

### PR DESCRIPTION
**Pull request by IgorSusmelj**
Add the NNCLR model from the paper. The only difference is that the projection head only has 2 layers at the moment. Reached 0.8358 accuracy on the test set for the CIFAR10 benchmark. This is more or less on par with the normal SimCLR model. Will do further experiments to see whether there is an error in the implementation.

### CODE REVIEW

File: nn_with_simsiam.py [hunk 1]
```diff
@@ -6,9 +6,10 @@
 import lightly
```
Comment (MalteEbner) [inline nn_with_simsiam.py:6]: The file should be renamed to `main.py` or `nnclr_on_cifar10` or similar.

File: nn_with_simsiam.py [hunk 2]
```diff
@@ -209,8 +208,40 @@ def configure_optimizers(self):
         scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optim, max_epochs)
         return [optim], [scheduler]
 
+class NNCLRModel(BenchmarkModule):
```
Comment (MalteEbner) [inline nn_with_simsiam.py:210]: Is there a reason why this class is not in the file `nnclr_model.py`? Furthermore, I do not get the difference between the two classes `NNCLRModel` and 'NNCLR'.

If the NNCLRModel truly only is the CLRModel, but with a different training step, could it inherit from it and only overwrite the training step? This would save a lot of code.

File: nnclr_model.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
+""" NNCLR Model """
```
Comment (MalteEbner) [inline nnclr_model.py:1]: I do not see any code about the nearest neighbour in this file, thus I do not really understand why it is called NNCLR. Isn't it just a slightly different implementation of CLR?

File: source/nn_memory_bank.py [hunk 4]
```diff
@@ -9,16 +9,15 @@ def __init__(self, size: int = 2 ** 16):
...
-        if normalize:
-            output = torch.nn.functional.normalize(output, dim=1)
+                update: bool = False):
 
         output, bank = super(NNmemoryBankModule, self).forward(output, labels, update)
         bank = bank.to(output.device).t()
 
-        similarity_matrix = torch.einsum("nd,md->nm", output, bank)
+        output_normed = torch.nn.functional.normalize(output, dim=1)
+        bank_normed = torch.nn.functional.normalize(bank, dim=1)
```
Comment (MalteEbner) [inline source/nn_memory_bank.py:18]: Is there a reason for moving the normalisation after the memory bank instead of before it? The memory bank is already initialised normally, thus this should not be the reason. This implementation of normalising the bank every time is more compute-intensive.

Comment (IgorSusmelj) [reply source/nn_memory_bank.py:18]: I just didn't want to enforce the normalization of the returned values.

**Review by MalteEbner — APPROVED** [review]
No major issues, let's have the tutorial to be able to push it.
